### PR TITLE
Drawing custom menu items for Menu Bar **Dark Theme Only**

### PR DESCRIPTION
**Pull request by ShahzaibIbrahim**
Using gc to draw menu item, moving the responsibility from OS due to wrong scaling of menu bar horizontally in **Dark** Theme. Snippet 373 can be used to test menuItems with images. 

Additional points covered in this PR: 

- Changing the text color from a grayish tone to white.
- When ALT key is pressed mnemonics are underlined and work as a toggle, the behavior which was missing from dark theme previously.

Screenshot of the fix: 

image

## HOW TO TEST

- Run the `RunTimeWorkspace` with the following **VM Arguments**
```
-Dswt.autoScale=quarter
-Dswt.autoScale.updateOnRuntime=true
```
- Go to "Windows" -> "Preferences" -> "General" -> "Appearance" 
- Change theme to "Dark" 
- Click "Apply and Close" 
- Move the window from 100 -> 175 (or anything 100+) zoom level
- See if menu bar is properly aligned.

## SCENARIOS TESTED

autoscale.updateOnRuntime = false && autoscale=int200 (Prevent regressions)
100% (Primary), 175% (Second)
150% (Primary), 200% (Second)
175% (Primary), 125% (Second)
200% (Primary), 100% (Second)
200% (Primary), 150% (Second)
autoscale.updateOnRuntime = false && autoscale=quarter (Prevent regressions)
150% (Primary), 200% (Second)
175% (Primary), 125% (Second)
autoscale.updateOnRuntime = true && autoscale=int200
100% (Primary), 175% (Second)
150% (Primary), 200% (Second)
175% (Primary), 125% (Second)
200% (Primary), 100% (Second)
autoscale.updateOnRuntime = true && autoscale=quarter
150% (Primary), 200% (Second) 
175% (Primary), 125% (Second)

## EXPECTED BEHAVIOUR

The menu bar should be scaled properly.

contributes to https://github.com/eclipse-platform/eclipse.platform.swt/issues/62 and https://github.com/eclipse-platform/eclipse.platform.swt/issues/127

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1164]: This has set the left margin of an entry before. now the margin appears to be zero and has to be added at some point. The first element is drawn exactly at x coordinate 0, which looks a bit odd:
Screenshot 2024-08-19 130842

Here is a comparison with the existing spacing (on top). Not only the initial left margin is different, but also the other spacing seems to be different (larger than before):
Screenshot 2024-08-19 130854

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1142]: While in general I like the new colors more than the old ones (in particular the white instead of the gray text in a selected entry), the region that is colored is incorrect. Only the label is affected, but actually the background of the whole menu item should be drawn in that color, as it was done before. Maybe you can simply fill the whole menu item with that color first? And maybe it is also possible to use the old blue tone for that?
Here is how a selected item looks now:
Screenshot 2024-08-19 130943

In here is how it looked before:
Screenshot 2024-08-19 130933

You can also see that the text color has changed in general (from a grayish tone to white), which I personally prefer, but we have to consider that this is an additional change.

**Review by HeikoKlare — COMMENTED** [review]
I tested the changes with different scale factors and auto-scale settings (quarter/integer200). The scaling results look fine, in constrast to the existing scaling behavior that was broken.

I have some detailed comments and a general question: are there snippets to test the behavior also in combination with images? Our default applications only have menus with text but it would also be interested to test the changes with images, so we should either have or provide a snippet with images for validation.

Please also improve the commit message according to the [recommendations](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations), such that it explains the existing issue and more precisely describes the fix. E.g., it currently sounds as if the menu is always draw in a custom way by this change, but it only affects the behavior when using dark theme.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java:648]: Adding a new system color for require adaptations of every OS implementation (i.e., also in the `Device` classes of MacOS and Linux). I am not sure whether we should really add a color for this use case. In particular, the color does not seem to have a specific meaning. In my Windows installation, the color currently used for the menu bar in dark mode does not correspond to this value, i.e., it may also depend on the OS version. Since it does not fit to the hightlight color of menu items either (see screenshot below) and since in light theme there is no similar hightlight color used as well (also see screenshot below).

image
image

I wonder whether we should not better just use one of the already defined color that we consider fitting. Since the hovering is also confirmed by opening the menu, a good visible color is not as important as, e.g., for menu items, where you want to see over which you currently hover. What about, for example, just using `COLOR_DARK_GRAY`? That would look as follows.
image

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1141]: Please extract this into a static constant and give it a proper, self-explaining name that makes the comment obsolete (like `leftTextMargin`).

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1146]: Please fix broken formatting (this marker is only made exemplarily)

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1158]: This comment is misplaced. Preivously, it belonged to a different line of code.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1163]: Please enclose blocks in braces.

**Review by HeikoKlare — CHANGES_REQUESTED** [review]
Thank you for the recent improvements. There are several improvements in the new version, that's great. In particular, this change also improves the visualization of menus with images. This is how it looked before:
image
And this is how it looks now:
image
Only problem is that the left margin is always applied to the text, even if an image is present. But in that case, the margin should be applied to the image (and maybe some additional small margin needs to be placed beetween image and text). This is, e.g., how it looks for the light theme:
image

One additional issue I found concerns the handling of mnemonics, i.e., when pressing Alt+$ for accesing menu item with menmonic $.
1. The visualization is broken: the background color of the menu items does not fit when pressing the "Alt" button (it is black while the rest of the menu is grayish). See the screenshot below.
2. Mnemonics are always displayed (underlined) and not only when Alt is pressed. This is, actually, existing behvior, but differs from default behavior, such as in light theme, where the mnemonics are only underlined when Alt is pressed. Maybe we could address this as a follow up, as no regression is introduced. On the contrary, in the current behavior, pressing Alt hides the mnemonics, so badly inverts the intended behavior.
3. Most severe issue: there is no reaction to key bindings for mnemonics. E.g., pressing Alt+F should open the "File" menu, but does nothing.

image

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java
Comment (akurtakov) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: If you introduce such new API you shoud do it for Cocoa and GTK too.

Comment (ShahzaibIbrahim) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Device.java]: We are now sticking to the default color that is already present i.e. COLOR_DARK_GRAY

**Comment by HeikoKlare** [comment]
Something seems to be wrong with the PR now. No matter whether I rebase on #1349, I see two issues:
1. The left margin of each menu item is quite large 
image
2. There are unexpected menu items, like another "File" entry with only a single entry in the menu that opens up from it 
image

Can you please check that, @ShahzaibIbrahim?

**Comment by HeikoKlare** [comment]
I retested and found that the faulty second "File" entry is because of a broken workbench.xmi on my side. Still, it shows some placement issue, as the text is cut off (hapenning for a "File" entry without a mnemonic, while the original "File" entry with mnemonic is not cut off). That may be investigated. You can easily reproduce by manually adding two menus (with at least one dummy menu item in each of them to have the menu shown) through the Model Spy:
image

**Comment by HeikoKlare** [comment]
Please check your latest commit. Only considering the diff, it does not seem to be in the intended/final state. There is a `System.out.println()` statement and some some added magic number in addition to the essential change:
image

**Comment by ShahzaibIbrahim** [comment]
> Please check your latest commit. Only considering the diff, it does not seem to be in the intended/final state. There is a `System.out.println()` statement and some some added magic number in addition to the essential change: image

Yes sorry, that was a leftover.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1163]: Isn't this relevant anymore, so that the width can be set to a different value?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1182]: Would be good to put this into a self-explaining method, such as `calculateRenderedTextSize()`.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:51]: How are these values derived? The margins look rather different to the existing implementation and to the light theme. In the existing implementation, a value of "5" was hardcoded for the left margin at some point, maybe that would fit as well?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1154]: In addition, it seems to be incorrect to scale down the predefined values, as they are the desired, consistent values in points, aren't they? With the current implementations, the margin are inconsistent across different scale values (I would say half the size on 200% compared to 100%). This applies to all scaleDown operations in this change including constant values in points.

Example 100%:
image

Example 175%:
image

Note that the _left_ margin on 175% is near the expected one while the one at 100% is way to high (compared with existing behavior / light theme). The overall margins seem to be too large in both cases.

**Review by HeikoKlare — COMMENTED** [review]
The proposed change works really great now! It fixes the existing issues with scaling the menu bar in dark theme and it also improves the appearance of the menu bar in dark theme overall.

I only have one behavioral concern left, which is about the statically defined margins that (1) seem to be inappropriate and (2) are scaled incorrectly. The other comments are regarding code style / design and rather nitpicky.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (ShahzaibIbrahim) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1152]: I have now refactored and assigned them to more self-explanatory variables and then passed menuItemArea for fillRectangle. As for `bounds.width` I have refactored its name `menuItemBounds `to be more clear.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (ShahzaibIbrahim) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1163]: This information still seems relevant on explaining the difference between `left margin` and `item size`. Also we are back to using value of 5 again to match the margins between items in menu bar. I would suggest to keep this comment as is.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (ShahzaibIbrahim) [reply bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:51]: I am going back to the value 5 for item width as before (because of larger gap) and left margin value was derived as a result of testing against different values and finding out the closest to the light theme.

**Review by HeikoKlare — COMMENTED** [review]
Code quality improvements are overall fine. The behavioral concerns in the last review have not been addressed yet. Please check them again.

In addition, testing Snippet 373 shows a misplacement of the image: 
image
image

One minor finding that may be addressed independently as it is not part of this PR:
the image is not scaled correctly when using `rescaleOnRuntime`:
image

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1136]: Just saw that this comment still seems to be relevant, so would be good to revert its removal.

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1156]:
```suggestion
			gc.setBackground(isSelected ? display.getSystemColor(SWT.COLOR_DARK_GRAY) : parent.getBackground());
```

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1159]: Double empty line seems unintended.
```suggestion

```

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1160]: Please fix the typo in `xPosiitonText`.

**Review by HeikoKlare — COMMENTED** [review]
Looks good now, thank you! Only some comments on documentation, formatting, typo left.

Could you please squash the commits into a single one?

**Review by HeikoKlare — APPROVED** [review]
This look good now, thank you!

Build failures are unrelated. Still we need to wait for them to be fixed until we can merge.

**Review by HeikoKlare — COMMENTED** [review]
Unfortunately, when doing another final testing after rebasing, I found a further issue. To reproduce, do the following:
* Have a primary monitor with >100% scaling (e.g., 125%), have a secondary monitor with 100% scaling
* Start an application with PerMonitorV2 DPI awareness in dark mode
* Move the application from 125% monitor (on which it looks fine) to 100% monitor.

The sizes seem to be calculated properly, but the font is too small (probably 100% instead of 125%):
{5A84A883-8FE7-4D82-A6BC-CFAE977C2ED5}

@akoch-yatta @ShahzaibIbrahim can you please have a look? Maybe this is easy to resolve by using proper scale values at some place.

**Review by HeikoKlare — COMMENTED** [review]
Unfortunately, it looks still almost the same. On 125% it became better/good, but with the primary monitor on 150% or above, I still see the issue. E.g., this is with primary monitor 175% on secondary monitor with 100%:
{A059FDB7-CBCC-4CDD-9505-CFCA3476717E}

**Comment by akoch-yatta** [comment]
> Unfortunately, it looks still almost the same. On 125% it became better/good, but with the primary monitor on 150% or above, I still see the issue. E.g., this is with primary monitor 175% on secondary monitor with 100%: {A059FDB7-CBCC-4CDD-9505-CFCA3476717E}

@HeikoKlare Sorry for all that back and forth, we need to improve the test protocol on our side, so you don't have that much work retesting changes. It is not yet properly working on all scenarios, as you mentioned. We'll have a look on that tomorrow.

**Comment by akoch-yatta** [comment]
In each of the following screenshot there is an adapted version of Snippet373 started with commit [2fe01bf](https://github.com/eclipse-platform/eclipse.platform.swt/commit/2fe01bfb14ea7ceb50c029267a073a0b06ff3ff9), 4 times each. The title gives the info of the configuration:

1. autoscaleOnRuntime active or not
2. autoscale mode: int200 or quarter 

-------------------------------------------------
Setup: Primary 125%, secondary 175%
Primary:
image
Secondary:
image

-------------------------------------------------
Setup: Primary 175%, secondary 125%
Primary:
image
Secondary:
image

-------------------------------------------------
Setup: Primary 150%, secondary 200%
Primary:
image
Secondary:
image

-------------------------------------------------
Setup: Primary 200%, secondary 100%
Primary:
image
Secondary:
image

**Comment by HeikoKlare** [comment]
Is this supposed to be ready for another review? I have only taken a quick look into an SDK application, but spacing is still off. See this screenshot at 150% (top one is with current implementation, bottom one with this PR):
{124AEFF1-853E-4512-8926-1BB6DCFECDCC}

**Comment by akoch-yatta** [comment]
Hmm, I cannot reproduce any vertical spacing issue, neither with quarter nor with int200. 
100% 
image
150% 
image
175%
image
200%
image

Regarding the horizontal indentation: I agree, that int direct comparision positioning could be improved. I tweaked some statics in https://github.com/eclipse-platform/eclipse.platform.swt/pull/1405/commits/cd3bb9bd972b495995e1e4757f38db37d79afadc to achieve the following result
100%
image
150%
image
200%
image

**Review by fedejeanne — APPROVED** [review]
The changes look fine to me and I didn't find any regressions/bugs while testing. 

I only have 2 nit-picky comments regarding readability but aside from that, this PR is good to go on my account.

**Review by fedejeanne — APPROVED** [review]
@ShahzaibIbrahim please squash your commits and this one is good to go on my account. @HeikoKlare should I merge or do you want to take another look?

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1192]:
```suggestion
			int yPositionText = DPIUtil.scaleDown(struct.top, zoom) + MARGIN_HEIGHT;
```

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java
Comment (HeikoKlare) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/MenuItem.java:1197]:
```suggestion
			int gap = (menuItemArea.height - image.getBounds().height) / 2;
```

**Review by HeikoKlare — APPROVED** [review]
I've retested this and did not find any issues anymore. In particular with existing HiDPI support (i.e., without per-monitor rescaling at runtime) and all combinations of settings (integer200 or quarter scaling, DPI awareness set to PerMonitorV2 or System, two monitors with different scalings between 100% and 200%).

The general appearance is also better than with the existing implementation (placement/spacing and highlight color). There is still some room for slight improvements on vertical alignment (with 100% is slightly off to the top while at 175% it is slightly off to the bottom), but that's rather nitpicky and, as said, the proposed state is still far better than the existing one.

@ShahzaibIbrahim can you please check the code formatting again (see, e.g., the comments I made) and squash the commits, so that we can merge this for M2?

**Review by fedejeanne — APPROVED** [review]

**Comment by HeikoKlare** [comment]
> Is there a shortage?

https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/5131

**Comment by HeikoKlare** [comment]
Jenkins is still unavailable, but (Jenkins) builds before latest update were successful and without new issues. Since [only formatting changes have been performed since then](https://github.com/eclipse-platform/eclipse.platform.swt/compare/372338d1058469845da8f9610f5b569554ca45bd..885fc638794b03091866673603f3546be0dd2692), this is still safe to merge.

**Pull request merged**
